File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    if target not in words:
        return 0
    
    queue = deque([(begin, 0)])
    visited = set([begin])
    while queue:
        current_word, cnt=queue.popleft()
        
        if current_word == target:
            return cnt
        
        for word in words:
            if word not in visited:
                diff_cnt=0
                for c1, c2 in zip(current_word, word):
                    if c1 != c2:
                        diff_cnt+=1
                if diff_cnt == 1:
                    visited.add(word)
                    queue.append((word, cnt+1))
    
    
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;
vector<bool> visited;
int answer;

void dfs(string curr, string target, vector<string>& words, int depth){
    if(curr==target) {
        answer=min(answer, depth);
        return;
    }
    for(int i=0; i<words.size(); i++){
        if(!visited[i]){
            int cnt=0;
            for(int j=0; j<target.size(); j++){
                if(curr[j]!=words[i][j]){
                    cnt++;
                }
            }
            if(cnt==1){
                visited[i]=true;
                dfs(words[i], target, words, depth+1);
                visited[i]=false;
            }
        }
    }
}


int solution(string begin, string target, vector<string> words) {
    visited.assign(words.size(), false);
    answer = 1e9;
    dfs(begin, target, words,0);
    if(answer==1e9){
        return 0;
    }
    return answer;
}